### PR TITLE
[historical-view] Allow empty minimized state from app-backend

### PR DESCRIPTION
**Pull request by Kongkille**
This allows ingestion of /state and /history with empty zonedata arrays

In the case of no data, we will have an empty overviews array, which also is the initial data state for the zone. 

Now dataReducer expects /history to return a bool hasData instead of checking in the frontend.

This seems to work, however, I expect to follow up with a PR where i deprecate hasData, hasParser and instead opt for a state status. This makes it a bit easier to incorporate construct-breakdown zones.

### CODE REVIEW

**Comment by VIKTORVAV99** [comment]
@Kongkille would this allow for some of the potential changes discussed in #4049 or is it just for empty zones?

**Comment by Kongkille** [comment]
> @Kongkille would this allow for some of the potential changes discussed in #4049 or is it just for empty zones?

Absolutely. The minimized state and history files I've tested so far have both removed keys with null values, and if there's no data at all we return an empty array.

That reminds me that I should update our mock data files to illustrate!

File: web/src/layout/map.jsx [hunk 1]
```diff
@@ -128,10 +128,9 @@ export default () => {
     () => (zoneId) => {
       const zoneOverview = zones[zoneId][selectedTimeAggregate].overviews[selectedZoneTimeIndex];
       const zoneConfig = zones[zoneId].config;
-      if (!zoneOverview || !zoneConfig) {
-        return;
+      if (zoneOverview) {
+        dispatchApplication('co2ColorbarValue', getCO2IntensityByMode(zoneOverview, electricityMixMode));
       }
-      dispatchApplication('co2ColorbarValue', getCO2IntensityByMode(zoneOverview, electricityMixMode));
```
Comment (madsnedergaard) [inline web/src/layout/map.jsx:134]: Don't we still need the zoneConfig check or is that never null? :)

**Review by madsnedergaard — COMMENTED** [review]
Nice, haven't tested, but code seems fine :) What's missing from this PR going in?

File: web/src/layout/map.jsx (hunk 1 above)
Comment (Kongkille) [reply web/src/layout/map.jsx:134]: We should never have a zoneOverview and no config, but since that's the case I might as well add it for readability

**Comment by VIKTORVAV99** [comment]
While you guys are already making changes to the state files would it be possible to add a function that rounds all values to an appropriate amount of decimals so we don't end up with these? `"fossilFuelRatioProduction":0.18719999999999998`

I think 2 decimals should be enough on the frontend but you might want to include 1 or 2 extra (3 and 4 total respectively) just in case.

Obviously this isn't required and shouldn't take priority but it would be nice to have.

**Review by madsnedergaard — COMMENTED** [review]
A few minor comments, otherwise awesome work! 💪

**Comment by VIKTORVAV99** [comment]
> @VIKTORVAV99 Ah good point on rounding. I've made an issue internally for the app-backend.
> 
> We still have a few nulls on production mix, which arguably could be removed, but I think it's OK for now, as the primary concern was the state files.
> 
> As for exchanges, in the spirit of time, we have not looked further into exchanges just yet. This will be one of the first things we will look into after launch, right now exchanges will only be available for the current hour in hourly time.
> 
> I will have to park it here for now but I'd be happy to look further into this after launch!

It should be relatively safe to just drop the exchanges with null netFLow from the state files all together as they won't show up on the map anyway. And if I'm not mistaken the history files will be the ones that are used for the exchange data in the "country details" page in the left panel.

Regarding the production mix I say take your time, it is not required and should not be a priority by any means. Just add it to the backlog 😉

File: web/cypress/integration/ranking.spec.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ describe('Ranking Panel', () => {
 
     // Click a country and return the the ranking panel
     cy.get('.zone-list a.selected').click();
-    cy.contains('Data is temporarily unavailable for the selected time');
+    cy.get('.country-name').should('exist');
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/cypress/integration/ranking.spec.js:16]: Oh nice catch haha

File: web/src/helpers/redux.js [hunk 3]
```diff
@@ -39,9 +39,12 @@ function initDataState() {
     zone.config.capacity = zoneConfig.capacity;
     zone.config.contributors = zoneConfig.contributors;
     zone.config.timezone = zoneConfig.timezone;
-    zone.config.hasParser = (zoneConfig.parsers || {}).production !== undefined;
+    // hasParser is true if parser exists, or if estimation method exists
+    zone.config.hasParser =
+      (zoneConfig.parsers || {}).production !== undefined || zoneConfig.estimation_method !== undefined;
     zone.config.delays = zoneConfig.delays;
     zone.config.disclaimer = zoneConfig.disclaimer;
+    zone.config.countryCode = key;
```
[the remark this answers is not in the document]
Comment (Kongkille) [reply web/src/helpers/redux.js:47]: Yes well, there were some places where we depended on countryCode being part of the API response.

There may be more elegant ways to extract this as we should be able to always find it given it's the accessor key for the zone, but right now it works well with my 'combineData' method in /helpers/redux.

image

**Comment by Kongkille** [comment]
> It should be relatively safe to just drop the exchanges with null netFLow from the state files all together as they won't show up on the map anyway.

I agree that it probably doesn't require much, but I'm still a bit hesitant to do it right now as exchanges is a bit of a box to open, and we're trying to wrap up for a first version to be somewhat ready this week 😬

**Comment by VIKTORVAV99** [comment]
> > It should be relatively safe to just drop the exchanges with null netFLow from the state files all together as they won't show up on the map anyway.
> 
> I agree that it probably doesn't require much, but I'm still a bit hesitant to do it right now as exchanges is a bit of a box to open, and we're trying to wrap up for a first version to be somewhat ready this week 😬 

No worries!
Do what you feel is best right now and loop back once you guys have time.
I know you guys have a lot of work going on right now and this doesn't really do anything except reduce the file size a little so it's far from a priority IMO. (but should still be done eventually!) 

I'm looking forward to seeing the historic view in action!

**Pull request merged**